operate/importer-common/src/main/java/io/camunda/operate/zeebeimport/cache/FlowNodeInstanceTreePathCache.java

```diff
@@ -0,0 +1,144 @@
+   * as it corresponds to a FNI on root level.
+   *
+   * <p>When the cache doesn't contain the treePath for a given flowScopeKey, the corresponding
+   * treePathResolver is used (specified on cache construction).
+   *
+   * <p>Does the resolver has also no knowledge about the treePath the process instance key is used.
+   *
+   * @param compositeKey a composite key that contains information used to resolve the tree path for
+   *     the flow node instance
+   * @return the treePath of the flow node instance
+   * @throws IllegalArgumentException when the flow node instance record doesn't correspond to a
+   *     supported partition
+   */
+  public String resolveTreePath(final FNITreePathCacheCompositeKey compositeKey) {
+    final var partitionCache = backedTreePathCache.get(compositeKey.partitionId());
+    if (partitionCache == null) {
+      final IllegalArgumentException illegalArgumentException =
+          new IllegalArgumentException(
+              String.format(
+                  "Expected to find treePath cache for partitionId %d, but found nothing. Possible partition Ids are: '%s'.",
+                  compositeKey.partitionId(), backedTreePathCache.keySet()));
+
+      LOGGER.error(
+          "Couldn't resolve tree path for given partition id {}",
+          compositeKey.partitionId(),
+          illegalArgumentException);
+      throw illegalArgumentException;
+    }
+
+    return resolveTreePath(partitionCache, compositeKey);
+  }
+
+  private String resolveTreePath(
+      final Map<String, String> partitionCache, final FNITreePathCacheCompositeKey compositeKey) {
+    String parentTreePath;
+    // if scopeKey differs from processInstanceKey, then it's inner tree level and we need to search
+    // for parent 1st
+    if (compositeKey.flowScopeKey() == compositeKey.processInstanceKey()) {
+      parentTreePath = ConversionUtils.toStringOrNull(compositeKey.processInstanceKey());
+    } else {
+      // find parent flow node instance
+      parentTreePath =
+          partitionCache.get(ConversionUtils.toStringOrNull(compositeKey.flowScopeKey()));
+
+      // cache miss: resolve tree path
+      if (parentTreePath == null) {
+        parentTreePath = treePathResolver.apply(compositeKey.flowScopeKey());
+        LOGGER.debug(
+            "Cache miss: resolved treePath {} for flowScopeKey {} via given resolver.",
+            parentTreePath,
+            compositeKey.flowScopeKey());
+
+        // add missing treePath to cache
+        if (parentTreePath != null) {
+          partitionCache.put(
+              ConversionUtils.toStringOrNull(compositeKey.flowScopeKey()), parentTreePath);
+        } else {
+          LOGGER.warn(
+              "Unable to find parent tree path for flow node instance id [{}], parent flow node instance id [{}]",
+              compositeKey.recordKey(),
+              compositeKey.flowScopeKey());
+          parentTreePath = ConversionUtils.toStringOrNull(compositeKey.processInstanceKey());
```

Comment:
Shouldn't this be put into the cache as well? (I mean, if there is no `treePath`, then the next one doesn't need to do a search request.)

Reply:
Here I was not sure, as I was also not sure whether the error handling itself is correct tbh. Because we didn;t found the flowScope and just assume it is the PI 🤔

Reply:
As this is the same as before I don't want to change as part of this PR. Happy to do it in an additional PR when we have clarified whether this is correct or not 😅